### PR TITLE
NIFI-13620 : Fixed QueryCassandra Max Wait Time issue

### PR DESCRIPTION
**Pull request by rahulsharma980**
-  Resolved the problem where QueryCassandra was not respecting the configured Max Wait Time.
-  Updated the logic to ensure that the Max Wait Time is correctly applied to all query executions.

### CODE REVIEW

File: nifi-nar-bundles/nifi-cassandra-bundle/nifi-cassandra-processors/src/main/java/org/apache/nifi/processors/cassandra/QueryCassandra.java [hunk 1]
```diff
@@ -277,7 +279,9 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
             final ResultSet resultSet;
 
             if (queryTimeout > 0) {
-                resultSet = connectionSession.execute(selectQuery, queryTimeout, TimeUnit.MILLISECONDS);
+                Statement statement = new SimpleStatement(selectQuery);
+                statement.setReadTimeoutMillis((int) queryTimeout);
```
Comment (exceptionfactory) [inline nifi-nar-bundles/nifi-cassandra-bundle/nifi-cassandra-processors/src/main/java/org/apache/nifi/processors/cassandra/QueryCassandra.java:283]: This cast to `(int)` is not necessary.
```suggestion
                statement.setReadTimeoutMillis(queryTimeout);
```

**Review by exceptionfactory — COMMENTED** [review]
Thanks for the correction to this behavior @rahulsharma980. I noted one issue regarding an unnecessary cast to `int`.

**Comment by rahulsharma980** [comment]
Updated !

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
@rahulsharma980 This change appears to have broken unit tests across all platforms, so please review the failure logs to ensure that all unit tests pass.

```
java.lang.AssertionError: java.lang.NullPointerException: Cannot invoke "com.datastax.driver.core.ResultSet.getColumnDefinitions()" because "rs" is null
	at org.apache.nifi.util.StandardProcessorTestRunner.run(StandardProcessorTestRunner.java:233)
	at org.apache.nifi.util.StandardProcessorTestRunner.run(StandardProcessorTestRunner.java:189)
	at org.apache.nifi.processors.cassandra.QueryCassandraTest.testProcessorELConfigJsonOutput(QueryCassandraTest.java:218)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
Caused by: java.lang.NullPointerException: Cannot invoke "com.datastax.driver.core.ResultSet.getColumnDefinitions()" because "rs" is null
	at org.apache.nifi.processors.cassandra.QueryCassandra.convertToJsonStream(QueryCassandra.java:572)
	at org.apache.nifi.processors.cassandra.QueryCassandra.convertToJsonStream(QueryCassandra.java:546)
	at org.apache.nifi.processors.cassandra.QueryCassandra.lambda$onTrigger$1(QueryCassandra.java:307)
	at org.apache.nifi.util.MockProcessSession.write(MockProcessSession.java:931)
	at org.apache.nifi.util.MockProcessSession.write(MockProcessSession.java:67)
	at org.apache.nifi.processors.cassandra.QueryCassandra.onTrigger(QueryCassandra.java:299)
	at org.apache.nifi.processor.AbstractProcessor.onTrigger(AbstractProcessor.java:27)
	at org.apache.nifi.util.StandardProcessorTestRunner$RunProcessor.call(StandardProcessorTestRunner.java:287)
	at org.apache.nifi.util.StandardProcessorTestRunner$RunProcessor.call(StandardProcessorTestRunner.java:281)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
	```

**Review by exceptionfactory — CHANGES_REQUESTED** [review]
@rahulsharma980, reverting the change to cast the timeout to an `int` does not address the test failure issue. The problem appears to be the mocked behavior in `QueryCassandraTest` needs to be changed.

**Comment by exceptionfactory** [comment]
@rahulsharma980 Are able to make the changes to the test class to correct the build failures? As mentioned previously, this Processor is deprecated, and removed from the main branch. With that in mind, this pull request could still correct the behavior on the support branch for a potential subsequent release, but the automated tests must pass in order to move forward.

**Comment by rahulsharma980** [comment]
Hello Apache NiFi Team,

The branch you mentioned (support/nifi-1.x) is encountering build failures
due to a dependency issue, even without any changes, as shown in the
attached screenshot.
[image: image.png]

Additionally, the build is also failing for the Cassandra bundle due to a
similar dependency issue. Please see the attached screenshot for more
details.
[image: image.png]

Could you please advise on how to resolve these dependency issues?

Regards

**Comment by exceptionfactory** [comment]
@rahulsharma980 Unfortunately the referenced images did not come through in the comment. Can you include the text of the error in another comment?

**Comment by rahulsharma980** [comment]
Hello Apache NiFi Team,

The branch you mentioned (support/nifi-1.x) is encountering build failures
due to a dependency issue, even without any changes, Below is the error.

[ERROR] Failed to execute goal on project nifi-web-api: Could not resolve
dependencies for project org.apache.nifi:nifi-web-api:war:1.28.0-SNAPSHOT:
Could not find artifact
org.apache.nifi:nifi-web-security:jar:tests:1.28.0-SNAPSHOT in
apache.snapshots (https://repository.apache.org/snapshots)

Additionally, the build is also failing for the Cassandra bundle due to a
similar dependency issue. Below is the error.

[ERROR] Failed to execute goal on project nifi-cassandra-services: Could
not resolve dependencies for project
org.apache.nifi:nifi-cassandra-services:jar:1.28.0-SNAPSHOT: Could not find
artifact org.apache.nifi:nifi-ssl-context-service-api:jar:1.28.0-SNAPSHOT
in apache.snapshots (https://repository.apache.org/snapshots)


Could you please advise on how to resolve these dependency issues?

Regards

**Comment by exceptionfactory** [comment]
@rahulsharma980 Based on the error, it sounds like you are attempting to build just the individual module. It is necessary to build from the project root directory, although it is possible to limit the build to particular modules using the `-pl` argument for Maven.

**Comment by exceptionfactory** [comment]
@rahulsharma980 Given that this applies to a Processor that is now deprecated and removed from the main branch, it is not a primary focus at this time. If you are able to resolve the automated build failures with correct tests, we could reconsider this pull request. It is possible to run the GitHub actions in your fork, so I recommend getting a clean build there before opening the pull request again.

**Pull request closed without merging**